Issue 63: Implementing SBC (#68)
CDCgov/forecasttools-py: notebooks/sbc_model_checking.qmd

Reviewer: * Verifying (2025-02-11) on this that I was able to run everything successfully, without error or warning, and that I added the plots as a comment on this PR.
* My preference (free to disregard, here) would be for the section-titles to have the first letter of each word capitalized.
* Wikipedia links added to certain terms here might be useful but also might not be worth the time.

---
title: "Using the `SBC` class to do sim-based calibration for `numpyro` models"
format: gfm
engine: jupyter
execute:
    warning: false
    output: false
    cache: true
bibliography: references.bib
---

This notebook covers:

-   A brief introduction to simulation-based calibration (SBC) as a method for testing the self-consistency of a Bayesian inference method.
-   Tuning and running SBC for a `numpyro` model using the `SBC` from `forecasttools`.
-   Plotting results and interpreting them.

> The example we will use to illustrate our ideas is the classic "eight schools" inference problem from section 5.5 of Gelman *et al* [@gelman2013] with posterior sampling done using the NUTS algorithm implemented by `numpyro` PPL.
> 

Reviewer: ```suggestion
The example we will use to illustrate our ideas is the classic "eight schools" inference problem from section 5.5 of Gelman *et al* [@gelman2013] with posterior sampling done using the NUTS algorithm implemented by `numpyro` PPL[^eight_schools_note].

[^eight_schools_note]: The interested reader can find a worked example of the eight schools problem on NumPyro's homepage (see [here](https://num.pyro.ai/en/stable/getting_started.html)), along with other examples, such as _Using NUTS for a simple hierarchical model_ (see [here](https://github.com/pyro-ppl/numpyro/blob/master/examples/baseball.py)). 
```

## Simulation-based calibration

> "Unit" testing the correctness of Bayesian probabilistic programs is challenging because the target quantity is a distribution (the posterior distribution) typically sampled from via a random process, e.g. Markov chain Monte Carlo sampling. The range of models where the correct posterior distribution is known analytically, and therefore, we can test sampling against distributional checks (e.g. Kolmogorov-Smirnov tests, chi-squared tests, etc) is limited.

Reviewer: ```suggestion
Unit testing the correctness of Bayesian probabilistic programs is challenging because the target quantity is a distribution (the posterior distribution), which is typically sampled via a random process, such as Markov chain Monte Carlo sampling. The range of models for which the correct posterior distribution is known analytically is limited, making it difficult to test sampling methods against distributional checks (e.g., Kolmogorov-Smirnov tests, chi-squared tests, etc.).
```

Had trouble reading this first time over. 

Perhaps the change (of which some of it is stylistic in nature) I've suggested is an improvement.

Simulation-based calibration (SBC) [@talts] aims to provide a correctness check for a Bayesian model and sampler using the self-consistency property of Bayesian inference:

$$
p(\theta) = \int p(y,\theta) dy = \int p(\theta | y) p(y) dy = \int p(\theta|y)\int p(y | \theta')p(\theta')d\theta' dy.
$$

In words, this states that the distribution of posterior samples from Bayesian inference gathered over many datasets $(y^{(i)})_{i=1, 2, \dots}$ is the prior distribution $p(\theta)$ *if* the datasets are generated using the prior distribution $y^{(i)} \sim p(y | \theta') p(\theta')$.

This self-consistency property of Bayesian inference provides a convenient method for checking if the model inference process $p(\theta | y)$ is performing as desired, since generating datasets $(y^{(i)})_{i=1, 2, \dots}$ from a generative model is typically quite easy. Since for any individual $k$th parameter $\theta[k] \in \theta$ we know the prior distribution $p(\theta[k])$ we can check correctness against:

$$
F^{(-1)}(\theta[k]) \sim \mathcal{U}(0,1).
$$

> Where $F^{(-1)}$ is the (pseudo) inverse distribution function of the known distribution $p(\theta[k])$. This is often call ed the probability integral transform (PIT). Significant (in a classical statistics sense) deviations in the PIT of parameters from the uniform distribution form evidence that either the model is incorrectly coded, its priors are cover hard to sample regimes, and/or the Bayesian inference method is approximate/invalid. See Talts *et al* [@talts] for more details on understanding the returned tests.

Reviewer: ```suggestion
Where $F^{(-1)}$ is the (pseudo) inverse distribution function of the known distribution $p(\theta[k])$. This is often called the probability integral transform (PIT). Significant (in a classical statistics sense) deviations in the PIT of parameters from the uniform distribution form evidence that either the model is incorrectly coded, its priors are cover hard to sample regimes, and/or the Bayesian inference method is approximate/invalid. See Talts *et al* [@talts] for more details on understanding the returned tests.
```

Grammar edit.
